### DISCUSSION TITLE
SoftRobots and SofaCUDA

### DISCUSSION THREAD

**Discussion by Gronum00** [SofaDefrost/SoftRobots#195] [Help - Programming with SoftRobots]
Greetings,

I am trying to use Cuda with SoftRobots and SofaPython3; which is also included in "all-included binaries (v22.06.00)" of SofaDefrost.
However it seems that it is not possible right away, and I cannot find any documentation online.

There are no problems when running the exaples .scn files for SofaCUDA and example .py scenes for SoftRobots separately.

I get errors for mainly for two parts(of SoftRobots plugin) in terminal when trying to edit and run a .py file scene. Which are related with mapping of both VisualModel and CableConstraint.

I am looking forward to hearing any answer, has anyone used these two plugins together before?
If these two plugins cannot be used together, is MultiThreading only alternative?

**Comment by alxbilger** [comment]
Which components from SoftRobots do you use in the Node where you have your Cuda Mechanical Object?

**Reply by Gronum00** [reply]
Thank you for the fast answer,

From SoftRobots, only "PullingCable"
From STLIB, "FixedBox", "CollisionMesh", "ElasticMaterialObject", "RigidObject", and lastly MainHeader and ContactHeader

I Modified STLIB functions/modes to work with cuda and when comment out all the cable related lines from the code, there is no error/warning in terminal but still the model is not animated (when animation button pressed nothing happens to the model, just time passes)

But I could not change VisualModel to CudaVisualModel in the code. Also When cables attached there is an error in the terminal that can be roughly interpreted as, It cannot map the PullingCable and the model ElasticMaterialObject

**Reply by alxbilger** [reply]
Do you have an error message in the log?

**Reply by Gronum00** [reply]
Again thank you for the attention.

Below, I copy paste the warning and error messages from terminal:


[WARNING] [CableConstraint(unnamed)] Requested template 'CudaVec3f' cannot be found in the list of available templates [Vec2d, Vec3d]. Falling back to default template: 'Vec3d'.
[INFO]    [SceneCheckerVisitor] Validating node "root" with checks: [SceneCheckDuplicatedName, SceneCheckMissingRequiredPlugin, SceneCheckUsingAlias, SceneCheckDeprecatedComponents, SceneCheckCollisionResponse]
[INFO]    [SceneCheckerVisitor] Finished validating node "root".
[ERROR]   [CableConstraint(CableConstraint)] There is no mechanical state associated with this node. The object is then deactivated. To remove this error message, fix your scene possibly by adding a MechanicalObject.
CUDA: BarycentricMapperMeshTopology: map initialized, 601 output points, 0 points ouside input mesh, max 4 inputs points per output, 2560 contributions total.



Note: Under "cable1" node (that contains CableConstraint which generates error) there is indeed a MechanicalObject whose template is CudaVec3f. In this cable node, there are 3 components: MechanicalObject, CableConstraint, Mapping

**Reply by alxbilger** [reply]
The disadvantage of SofaCUDA is that the MechanicalObject change its template parameter (from Vec3d to CudaVec3f). The consequence is that all components associated with this MechanicalObject, including CableConstraint, must be available with the template CudaVec3f. This is not the case for CableConstraint.

To make available some components for Cuda in the plugin SoftRobots, a file similar to https://github.com/sofa-framework/BeamAdapter/blob/master/src/BeamAdapter/gpu/cuda/CudaInstantiations.cpp must be written (and I don't guarantee it will work for a constraint).

How do you use SOFA? From the binaries or do you compile it yourself?

**Comment by alxbilger** [comment]
@epernod Did you try to combine SoftRobots and Cuda?

**Reply by epernod** [reply]
no sorry, I didn't use SofaRobots since a long time
